### PR TITLE
[System.IO.Compression] Brotli from CoreFX

### PR DESCRIPTION
**Pull request by EgorBo**
Resolves https://github.com/mono/mono/issues/11431

I put the native impl to Mono.Native, but probably it should be in a separate lib (`libmono-native.0.dylib` before: **81kb**, after: **898kb**) if Linker is not able to trim it if it's not used.
Compiles and works on Windows (however, we don't build Mono.Native for Windows, but we can borrow `clrcompression.dll` from .net core) and macOS but requires changes in mono/corefx,

I disabled xunit tests as those need 50mb of test data.

PS: Brotli could be fully managed 😢 https://github.com/dotnet/corefxlab/issues/1673

C# ports:

- https://github.com/master131/BrotliSharpLib 
A quick benchmark (Compressing a 5mb file with max compression level=11)
Mono JIT: 25sec
Mono LLVM-AOT: 7.8sec
.NET Core 2.2: 6.8sec
.NET Core 3.0: 5.8sec
.NET Core's BrotliStream (with native lib): 3.5sec

- https://github.com/google/brotli/tree/master/csharp - official C# port. Based on Java sources, slow, outdated (2 years ago).

### CODE REVIEW

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by luhenry** [comment]
@chamons @jonpryor heads up that this is going to impact XI/XA in the future, adding ~700kb of code.

**Comment by EgorBo** [comment]
That's why the code should be linked out in case if `BrotliStream` is not used in the app 🙂

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins commit csproj and apidiff

**Comment by marek-safar** [comment]
@monojenkins commit csproj and apidiff

File: mcs/build/tests.make [hunk 1]
```diff
@@ -92,7 +92,7 @@ endif
 tests_CLEAN_FILES += $(test_lib_output) $(test_lib_output:$(ASSEMBLY_EXT)=.pdb)
 
 xtest_sourcefile_base = $(PROFILE_PLATFORM)_$(PROFILE)_$(ASSEMBLY:$(ASSEMBLY_EXT)=_xtest.dll.sources)
-xtest_sourcefile_base_excludes = $(PROFILE)_$(ASSEMBLY:$(ASSEMBLY_EXT)=_xtest.dll.exclude.sources)
+xtest_sourcefile_base_excludes = $(PROFILE_PLATFORM)_$(PROFILE)_$(ASSEMBLY:$(ASSEMBLY_EXT)=_xtest.dll.exclude.sources)
```
Comment (marek-safar) [inline mcs/build/tests.make]: This breaks existing tests
```
13:59:27 CSC     [wasm] wasm_System_xunit-test.dll
13:59:27 /mnt/jenkins/workspace/test-mono-pull-request-wasm/external/corefx/src/System.CodeDom/tests/Compiler/CodeDomProviderTests.cs(9,17): error CS0234: The type or namespace name 'CSharp' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?)
```

**Comment by vargaz** [comment]
This breaks the ios sdk lane. Why was this merged ?

**Comment by EgorBo** [comment]
@vargaz oops, fixing now. Previously iOS lane failed with because of a config issue I fixed

File: mono/native/Makefile.am [hunk 2]
```diff
@@ -91,7 +114,7 @@ macos_sources += $(gss_sources)
 ios_sources += $(gss_sources)
 endif
 
-EXTRA_libmono_native_la_SOURCES = $(common_sources) $(macos_sources) $(ios_sources) $(linux_sources) $(aix_sources) $(unix_sources) $(gss_sources)
+EXTRA_libmono_native_la_SOURCES = $(common_sources) $(brotli_sources) $(macos_sources) $(ios_sources) $(linux_sources) $(aix_sources) $(unix_sources) $(gss_sources)
```
Comment (lambdageek) [inline mono/native/Makefile.am]: `brotli_sources` is never assigned any value.  It doesn't make sense to add it here.

File: mono/native/Makefile.am [hunk 3]
```diff
@@ -34,7 +34,30 @@ common_sources = \
 	../../external/corefx/src/Native/Unix/System.Native/pal_uid.c \
 	../../external/corefx/src/Native/Unix/System.Native/pal_uid.h \
 	../../external/corefx/src/Native/Unix/System.Native/pal_time.c \
-	../../external/corefx/src/Native/Unix/System.Native/pal_time.h
+	../../external/corefx/src/Native/Unix/System.Native/pal_time.h \
+	../../external/corefx/src/Native/AnyOS/brotli/tools/brotli.c \
```
Comment (lambdageek) [inline mono/native/Makefile.am]: This file exports `main` which broke the ios SDK lane.

**Comment by akoeplinger** [comment]
Fixed in b4c66a1d59298d64c75a8470e106847d1a9eb70f.

**Pull request merged**